pkg/kritis/cryptolib/verifier.go

```diff
@@ -49,13 +52,47 @@ type PublicKey struct {
 // NewPublicKey creates a new PublicKey. `keyType` contains the type of the
 // public key, one of Pgp, Pkix or Jwt. `keyData` contains the raw key
 // material. `keyID` contains a unique identifier for the public key. For PGP,
-// this should be the OpenPGP RFC4880 V4 fingerprint of the key.
-func NewPublicKey(keyType KeyType, keyData []byte, keyID string) PublicKey {
-	return PublicKey{
+// this should be the OpenPGP RFC4880 V4 fingerprint of the key. For PKIX and
+// JWT, the ID should contain valid URI characters.
+func NewPublicKey(keyType KeyType, keyData []byte, keyID string) (*PublicKey, error) {
+	switch keyType {
+	case Pgp:
+		err := validatePgpKeyID(keyData, keyID)
+		if err != nil {
+			return nil, err
```

Comment:
In BinAuthz public doc: https://cloud.google.com/binary-authorization/docs/reference/rest/v1/projects.attestors#attestorpublickey
PGP key id would be overwritten by API generated ID. Should we follow this behavior instead of throwing an error?

Reply:
Done. Justification:

In Binauthz, the ID is overwritten during the creation of the `Attestor` resource. In Kritis, it would be ideal to replicate this behavior at `AttestationAuthority` creation time, but we do not have a management service.

However, by enforcing this behavior in NewPublicKey(), we can minimize false negative errors in VerifyAttestation(). For PGP, the ID of an attestation comes directly from the private key (see [here](https://github.com/grafeas/kritis/blob/257ac88f6544df5a89ae4a600811f41eb90dcec5/pkg/kritis/cryptolib/pgp.go#L87)). Introducing this behavior allows users to use public keys which can correctly verify attestations, even if its user-provided ID is incorrect.

Note: for PKIX/JWT, we will need the Signer implementations to conform the attestation ID as it is done here.